satellite/marketing/marketingweb/server.go

```diff
@@ -128,20 +170,73 @@ func (s *Server) parseTemplates() (err error) {
 	return nil
 }
 
+// timeConverter formats form time input as time.Time.
+func timeConverter(value string) reflect.Value {
+	v, err := time.Parse("2006-01-02", value)
+	if err != nil {
+		return reflect.Value{}
+	}
+	return reflect.ValueOf(v)
+}
+
+// formToStruct decodes POST form data into a new offer.
+func formToStruct(w http.ResponseWriter, req *http.Request) (o marketing.NewOffer, e error) {
+	err := req.ParseForm()
+	if err != nil {
+		return o, err
+	}
+
+	if err := decoder.Decode(&o, req.PostForm); err != nil {
+		return o, err
+	}
+	return o, nil
+}
+
+// CreateOffer handles requests to create new offers.
+func (s *Server) CreateOffer(w http.ResponseWriter, req *http.Request) {
+	o, err := formToStruct(w, req)
+	if err != nil {
+		s.log.Error("failed to convert form to struct", zap.Error(err))
+		s.serveInternalError(w, req, err)
+		return
+	}
+
+	o.Status = marketing.Active
+	reqType := mux.Vars(req)["offer_type"]
+
+	if reqType == "referral-offer" {
+		o.Type = marketing.Referral
+	} else {
+		o.Type = marketing.FreeCredit
+	}
+
+	if _, err := s.service.InsertNewOffer(context.Background(), &o); err != nil {
+		s.log.Error("failed to insert new offer", zap.Error(err))
+		s.serveInternalError(w, req, err)
+		return
+	}
+
+	http.Redirect(w, req, "/", http.StatusSeeOther)
+}
+
+// serveNotFound handles 404 errors and defaults to 500 if template parsing fails.
 func (s *Server) serveNotFound(w http.ResponseWriter, req *http.Request) {
 	w.WriteHeader(http.StatusNotFound)
 
 	err := s.templates.pageNotFound.ExecuteTemplate(w, "base", nil)
 	if err != nil {
 		s.log.Error("failed to execute template", zap.Error(err))
+		s.serveInternalError(w, req, err)
+		return
 	}
 }
 
-func (s *Server) serveInternalError(w http.ResponseWriter, req *http.Request) {
+// serveInternalError handles 500 errors and renders err to the internalErr template.
+func (s *Server) serveInternalError(w http.ResponseWriter, req *http.Request, e error) {
+
 	w.WriteHeader(http.StatusInternalServerError)
 
-	err := s.templates.internalError.ExecuteTemplate(w, "base", nil)
-	if err != nil {
+	if err := s.templates.internalError.ExecuteTemplate(w, "base", e); err != nil {
```

Comment:
the line is quite long so keep it on a separate line

Reply:
Do we really want to expose our internal error? If we open the endpoint at some point to the public with authentication then we would need to remove this. We still have the log with the appropriate error.

Reply:
The referral-admin portal isn't ever going to be open to the public. Only a couple admins on the marketing team who will use it to manage offers behind a VPN. The reason we are exposing the error is if someone from Marketing gets an error, they can file a Jira ticket and include the error message in the screenshot which will help us fix it for them. This is an internal service.